Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+      appendingSuffixTexts: suffixTexts,
+      withSourceLocationConverter: sourceLocationConverter
+    )
 
-    // Compute the set of highlight ranges that land on this line. These
-    // are column ranges, sorted in order of increasing starting column, and
-    // with overlapping ranges merged.
-    let highlightRanges: [Range<Int>] = annotatedLine.diagnostics.map {
-      $0.highlights
-    }.joined().compactMap { (highlight) -> Range<Int>? in
-      if highlight.root != Syntax(tree) {
-        return nil
-      }
+    // Calculate the ranges of line numbers that should be displayed, taking into account annotations and context.
+    let rangesToPrint = computeLinesToPrintRanges(from: annotatedSourceLines)
 
-      let startLoc = highlight.startLocation(converter: slc, afterLeadingTrivia: true)
-      let startLine = startLoc.line
+    // Accumulate the fully annotated source files here.
+    var annotatedSource = ""
 
-      // Find the starting column.
-      let startColumn: Int
-      if startLine < lineNumber {
-        startColumn = 1
-      } else if startLine == lineNumber {
-        startColumn = startLoc.column
-      } else {
-        return nil
-      }
+    // Keep track if a line missing char should be printed.
+    var hasLineBeenSkipped = false
 
-      // Find the ending column.
-      let endLoc = highlight.endLocation(converter: slc, afterTrailingTrivia: false)
-      let endLine = endLoc.line
+    // Maximum numbers of digits in line number mark.
+    let maxNumberOfDigits = String(annotatedSourceLines.count).count
 
-      let endColumn: Int
-      if endLine > lineNumber {
-        endColumn = annotatedLine.sourceString.count
-      } else if endLine == lineNumber {
-        endColumn = endLoc.column
-      } else {
-        return nil
+    for annotatedLine in annotatedSourceLines {
+      /// If the current line number is not within the 'rangesToPrint', we continue to the next line
+      /// and set the 'hasLineBeenSkipped' flag, so that we can later insert a marker for the skipped lines.
+      guard rangesToPrint.contains(where: { $0.contains(annotatedLine.sourceLineNumber) }) else {
+        hasLineBeenSkipped = true
+        continue
       }
 
-      if startColumn == endColumn {
-        return nil
+      /// Add a skipped line marker to indicate that lines have been omitted, but only under two conditions:
+      /// 1. Lines have actually been skipped ('hasLineBeenSkipped' is true).
+      /// 2. There's already annotated content in 'annotatedSource', meaning this isn't the start of the annotated lines.
+      /// This ensures that skipped line markers only appear between annotated lines.
+      if hasLineBeenSkipped && !annotatedSource.isEmpty {
+        let skippedLineMarker = diagnosticDecorator.decorateBufferOutline("┆")
+        let skippedLine = indentString + String(repeating: " ", count: maxNumberOfDigits) + " " + skippedLineMarker + "\n"
+        annotatedSource.append(skippedLine)
```

Comment:
Nit: I generally prefer `+=` to `append`. But feel free to leave as-is if you disagree. Same in a couple more places.

Reply:
I've replaced all occurrence of `append` with `+=`. Thanks.